lib/plug/builder.ex

```diff
@@ -179,29 +181,49 @@ defmodule Plug.Builder do
   @spec compile(Macro.Env.t, [{plug, Plug.opts, Macro.t}], Keyword.t) :: {Macro.t, Macro.t}
   def compile(env, pipeline, builder_opts) do
     conn = quote do: conn
-    {conn, Enum.reduce(pipeline, conn, &quote_plug(init_plug(&1), &2, env, builder_opts))}
+    init_mode = builder_opts[:init_mode] || :compile
+
+    unless init_mode in [:compile, :runtime] do
+      raise ArgumentError, """
+      invalid :init_mode when compiling #{inspect env.module}.
+
+      Supported values include :compile or :runtime. Got: #{inspect init_mode}
+      """
+    end
+
+    ast =
+      Enum.reduce(pipeline, conn, fn {plug, opts, guards}, acc ->
+        {plug, opts, guards}
+        |> init_plug(init_mode)
+        |> quote_plug(acc, env, builder_opts)
+      end)
+
+    {conn, ast}
   end
 
   # Initializes the options of a plug at compile time.
```

Comment:
Should this comment be updated?

Reply:
We can either drop the comment or make it `Initializes the options of a plug in the configured init_mode`

Reply:
Can you submit a PR? Thanks!